### DISCUSSION TITLE
Explain when to use which client configuration in Securing Applications and Services Guide

### DISCUSSION THREAD

**Discussion by thomasdarimont** [keycloak/keycloak#22334] [Ideas]
The Securing Applications and Services Guide in the documentation needs a section explaining when to use which client type/configuration.

I propose the following scenarios with configurations for OpenID Connect / OAuth 2 client integrations. 
Does anyone see some scenario that I'm missing (appart from the missing SAML examples)?

Note that I deliberately left out the implict flow and direct access grant.

----

1) Scenario:  Server-side Web Application with optional state on the backend.
App Type: Server-side Web App
Description: A server-side Web App that handles the OIDC callbacks on the server side. This application might use HTTP-sessions (managed in memory or an external session store) to store state on the backend or the Client (User-Agent) via HTTP-only secure (encrypted) cookies.
Client Type: Confidential Client
Flow: Standard Flow Enabled (Authorization Code Flow)
Options: PKCE with S256

2) Scenario:  Backend for Frontend (BFF) with hosted SPA client application.
App Type: Server-side Web App
Description: A server-side Web App that hosts a frontend SPA but handles the OIDC callbacks on the server side. All calls to other resource servers are tunneled via the backend. The backend enriches the resource server requests with an access token. This application might use HTTP sessions (managed in memory or an external session store) to store state on the backend or client-side via HTTP-only secure (encrypted) cookies.
Client Type: Confidential Client
Flow: Standard Flow Enabled (Authorization Code Flow)
Options: PKCE with S256

3) Scenario: SPA Frontend modeled that works without a dedicated backend.
App Type: Single Page App (SPA)
Description: A browser-based JavaScript application that manages the token handling on the client side.
Client Type: Public Client
Flow: Standard Flow Enabled (Authorization Code Flow)
Options: PKCE with S256

4) Scenario: A mobile app that can access resource servers.
App Type: Mobile App
Description: A mobile app that uses the System browser for sign-in. The mobile app manages tokens directly.
Client Type: Public Client
Flow: Standard Flow Enabled (Authorization Code Flow)
Options: PKCE with S256

5) Scenario: A desktop app that can access resource servers.
App Type: Desktop App / Native App
Description: A desktop application that uses the System browser for logging in.
Client Type: Public Client
Flow: Standard Flow Enabled (Authorization Code Flow)
Options: PKCE with S256
(Note for Windows Applications that are running in a Kerberos-enabled environment, one can also leverage Keycloaks Kerberos support to automatically sign in with the currently logged-in Windows user, without any user interaction)

6) Scenario: CLI App with Browser based login on the same machine
App Type: CLI App
Client Type: Public Client
Flow: Standard Flow Enabled (Authorization Code Flow)
Options: PKCE with S256

7) Scenario: CLI App with Browser based login potentially on a separate machine
App Type: CLI App
Client Type: Public Client
Flow: Device Flow Enabled (Authorization Code Flow)
Options: PKCE with S256

8) Scenario: Represent a Web API as resource-server
App type: Web API
Client Type: Confidential Client
Notes: A web API can optionally be represented as Client without any enabled flows.
If the web-api is modeled as Client, we can define role mappings for the Client and use that within other clients' role mappings. Additionally, we can emit the clientId in a token for another client as an allowed audience.
The client secret of the confidential client is not used in this scenario.

9) Scenario: Web API / Resource server with a Service Account.
Description: Web API / Resource server with Service Account, which can obtain tokens for the service and access other resource servers
App type: Web API
Client Type: Confidential Client
Flow: Service-Account Enabled (Client-Credentials Grant Flow)

10) Scenario: OIDC / OAuth Identity Provider Integration
Description: We represent the identity provider we want to integrate with as a dedicated client.
App type: IdP Client
Client Type: Confidential Client
Flow: Standard Flow Enabled (Authorization Code Flow)
Options: PKCE with S256
Notes: Some Identity Providers might not support configuration options, e.g., PKCE.

**Comment by danielFesenmeyer** [comment]
I think it would be really helpful to have such guidelines.
